### ISSUE TITLE
Handling arbitrary arrays in bincount

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#4857]
Currently we are [hardcoding a NumPy array in `bincount`]( https://github.com/dask/dask/blob/2f308e72a46de818055819108429e8f7ce7ba03b/dask/array/routines.py#L521 ). It would be good to match the result array type with the arrays provided as input.

**Comment by jakirkham** [comment]
cc @pentschev (in case this is of interest)

**Comment by jakirkham** [comment]
Just as one more note, here is the (shortened) traceback one sees when using CuPy arrays.

```python
~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/routines.py in _bincount_sum()
    520     out = np.zeros(n, dtype=dtype)
    521     for b in bincounts:
--> 522         out[:len(b)] += b
    523     return out
    524 

cupy/core/core.pyx in cupy.core.core.ndarray.__array_ufunc__()

cupy/core/_kernel.pyx in cupy.core._kernel.ufunc.__call__()

cupy/core/_kernel.pyx in cupy.core._kernel._preprocess_args()

TypeError: Unsupported type <class 'numpy.ndarray'>
```

**Comment by pentschev** [comment]
Unfortunately, this is not the only place there are hardcoded arrays. Ultimately, I think we will have to handle all existing calls to array creation functions, which include:

- `empty` and `empty_like` (as well as their counterparts, `full`, `ones`, `zeros`);
- `array` and `asarray`;
- possibly others I haven't seen or can't think of now.

**Comment by jakirkham** [comment]
At least for this case, we might be able to get away with something like...

1. Copy the larger of the two arrays.
2. Add the smaller array into the larger one.
3. Get the next array and repeat.

There's some inefficiency with this proposal, but that is true of ragged arrays in Python generally, which is what we are effectively working with in this case.

Not to mention we may want to adjust this code to leverage reductions as mentioned issue ( https://github.com/dask/dask/issues/4852 ), which may help by breaking this into smaller reductions that can be done in parallel.

Agree that we need to think a bit more about array creation generally. Is there already an open issue for this? If not, would you be ok adding one?

**Comment by pentschev** [comment]
Do you have a sample to reproduce the issue above @jakirkham? I think with `_meta` we can easily solve that by replacing the creation of `out` with `out = np.zeros_like(bincounts._meta, shape=n, dtype=dtype)`. This will require NumPy 1.17, but it's one of the cases we can't really deal without the new `shape` argument.

No, there's no open issue yet. I've been thinking of possible solutions and that's why I haven't raised an issue yet. Also, this is probably a broader discussion, that will require something in NumPy, so we can't solve it on Dask side alone. In fact, I've encountered the same issues in SciPy, so we definitely need a different way of creating arrays based on some input type.

**Comment by jakirkham** [comment]
Sure, here's a simple example.

```python
In [1]: import numpy as np                                                                            

In [2]: import cupy as cp                                                                             

In [3]: import dask.array as da                                                                       

In [4]: a = np.array([3, 0, 3, 9, 8, 9, 2, 0, 7, 6, 6, 7, 7, 0, 2, 5, 3, 8, 8, 0], dtype='u1')        

In [5]: d = da.from_array(a, chunks=3).map_blocks(cp.asarray)                                         

In [6]: r = da.bincount(d)                                                                            

In [7]: r.compute() 
```

**Comment by jakirkham** [comment]
> No, there's no open issue yet. I've been thinking of possible solutions and that's why I haven't raised an issue yet. Also, this is probably a broader discussion, that will require something in NumPy, so we can't solve it on Dask side alone. In fact, I've encountered the same issues in SciPy, so we definitely need a different way of creating arrays based on some input type.

It would be great if we raised the issue on Dask to start. Agree we might need to go in different directions from there, but that at least gives us a starting point to do that from.

**Comment by pentschev** [comment]
Sure, will formulate that nicely and raise an issue on Dask tomorrow to start. Thanks for the example, I'll check how to make sure it works with `_meta`.

**Comment by jakirkham** [comment]
Thanks Peter! 😄

**Comment by pentschev** [comment]
Glad to help! :)

**Comment by pentschev** [comment]
Fixed in https://github.com/dask/dask/pull/4543 now, but requires NumPy 1.17 (unreleased), and adding `shape` argument to CuPy `*_like` functions https://github.com/cupy/cupy/pull/2171, which is unmerged and won't be merged until NumPy 1.17 is released. I found also a bug with the CuPy PR I mentioned, which I fixed locally and will push a commit once I add tests for that.

**Comment by jsignell** [comment]
Seems like this one is done!
